### PR TITLE
Add `reserve` for `list`

### LINKED ISSUE #2232: Support `reserve` like `std::vector<T>::reserve` in C++
It looks like reserve makes a big difference here, so we should support it in LPython as well. For example by `from lpython import reserve` and just call this function. The CPython implementation will not do anything.

Here is how to use it:
```python
from lpython import reserve
def dijkstra_shortest_path(n: i32, source: i32) -> i32:
    i: i32; j: i32; v: i32; u: i32; mindist: i32; alt: i32; dummy: i32; uidx: i32
    dist_sum: i32;
    graph: list[i32] = []
    dist: list[i32] = []
    reserve(graph, n*n)
```

In CPython emulation this will be an empty function (no-op), in LPython we reserve the memory for this list, similar to reserve in C++, for performance.

_Originally posted by @certik in https://github.com/lcompilers/lpython/issues/2231#issuecomment-1657196590_

### PR DESCRIPTION
**Pull request by kabra1110**
Fixes #2232.

Similar to C++, I have kept this as a function intrinsic to list class.

The test does not work with CPython, as this function is not present there. I'm not entirely sure on how to add a no-op function for this to work.

### CODE REVIEW

File: integration_tests/test_list_reserve.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from lpython import i32, f64
+
+def test_list_reserve():
+    l1: list[i32] = []
+    l2: list[list[tuple[f64, str, tuple[i32, f64]]]] = []
+    i: i32
+
+    l1.reserve(100)
```
Comment (czgdp1807) [inline integration_tests/test_list_reserve.py:8]:
```suggestion
    reserve(l1, 100)
```

File: integration_tests/test_list_reserve.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    for i in range(50):
+        l1.append(i)
+        assert len(l1) == i + 1
+
+    l1.reserve(150)
```
Comment (czgdp1807) [inline integration_tests/test_list_reserve.py:13]:
```suggestion
    reserve(l1, 150)
```

File: integration_tests/test_list_reserve.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+    for i in range(50):
+        l1.pop(0)
+        assert len(l1) == 49 - i
+
+    l2.reserve(100)
```
Comment (czgdp1807) [inline integration_tests/test_list_reserve.py:19]:
```suggestion
    reserve(l2, 100)
```

File: integration_tests/test_list_reserve.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    for i in range(50):
+        l2.append([(f64(i * i), str(i), (i, f64(i + 1))), (f64(i), str(i), (i, f64(i)))])
+        assert len(l2) == i + 1
+
+    l2.reserve(150)
```
Comment (czgdp1807) [inline integration_tests/test_list_reserve.py:24]:
```suggestion
    reserve(l2, 150)
```

File: integration_tests/test_list_reserve.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+from lpython import i32, f64
```
Comment (czgdp1807) [inline integration_tests/test_list_reserve.py:1]:
```suggestion
from lpython import i32, f64, reserve
```

File: src/libasr/codegen/llvm_utils.h [hunk 6]
```diff
@@ -403,6 +403,9 @@ namespace LCompilers {
                             llvm::Module* module,
                             std::map<std::string, std::map<std::string, int>>& name2memidx);
 
+            void reserve(llvm::Value* list, llvm::Value* n,
+                         ASR::ttype_t* asr_type, llvm::Module* module);
+
```
Comment (czgdp1807) [inline src/libasr/codegen/llvm_utils.h:408]: This is exactly same as C++.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 7]
```diff
@@ -46,6 +46,7 @@ enum class IntrinsicFunctions : int64_t {
     Partition,
     ListReverse,
     ListPop,
+    ListReserve,
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:49]:
```suggestion
    Reserve,
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 8]
```diff
@@ -102,6 +103,7 @@ inline std::string get_intrinsic_name(int x) {
         INTRINSIC_NAME_CASE(Partition)
         INTRINSIC_NAME_CASE(ListReverse)
         INTRINSIC_NAME_CASE(ListPop)
+        INTRINSIC_NAME_CASE(ListReserve)
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:106]:
```suggestion
        INTRINSIC_NAME_CASE(Reserve)
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 9]
```diff
@@ -1262,6 +1264,52 @@ static inline ASR::asr_t* create_ListPop(Allocator& al, const Location& loc,
 
 } // namespace ListPop
 
+namespace ListReserve {
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:1267]:
```suggestion
namespace Reserve {
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 10]
```diff
@@ -1262,6 +1264,52 @@ static inline ASR::asr_t* create_ListPop(Allocator& al, const Location& loc,
...
+        x.base.base.loc, diagnostics);
+}
+
+static inline ASR::expr_t *eval_list_reserve(Allocator &/*al*/,
+    const Location &/*loc*/, Vec<ASR::expr_t*>& /*args*/) {
+    // TODO: To be implemented for ListConstant expression
+    return nullptr;
+}
+
+static inline ASR::asr_t* create_ListReserve(Allocator& al, const Location& loc,
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:1289]:
```suggestion
static inline ASR::asr_t* create_Reserve(Allocator& al, const Location& loc,
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 11]
```diff
@@ -1262,6 +1264,52 @@ static inline ASR::asr_t* create_ListPop(Allocator& al, const Location& loc,
...
+        x.base.base.loc, diagnostics);
+    ASRUtils::require_impl(ASR::is_a<ASR::Integer_t>(*ASRUtils::expr_type(x.m_args[1])),
+        "Argument to list.reserve must be an integer",
+        x.base.base.loc, diagnostics);
+    ASRUtils::require_impl(x.m_type == nullptr,
+        "Return type of list.reserve must be empty",
+        x.base.base.loc, diagnostics);
+}
+
+static inline ASR::expr_t *eval_list_reserve(Allocator &/*al*/,
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:1283]:
```suggestion
static inline ASR::expr_t *eval_reserve(Allocator &/*al*/,
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 12]
```diff
@@ -3124,6 +3172,8 @@ namespace IntrinsicFunctionRegistry {
             {nullptr, &DictValues::verify_args}},
         {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::ListPop),
             {nullptr, &ListPop::verify_args}},
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::ListReserve),
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:3175]:
```suggestion
        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Reserve),
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 13]
```diff
@@ -3206,6 +3256,8 @@ namespace IntrinsicFunctionRegistry {
             "list.reverse"},
         {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::ListPop),
             "list.pop"},
+        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::ListReserve),
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:3259]:
```suggestion
        {static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Reserve),
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 14]
```diff
@@ -3290,6 +3342,7 @@ namespace IntrinsicFunctionRegistry {
                 {"list.index", {&ListIndex::create_ListIndex, &ListIndex::eval_list_index}},
                 {"list.reverse", {&ListReverse::create_ListReverse, &ListReverse::eval_list_reverse}},
                 {"list.pop", {&ListPop::create_ListPop, &ListPop::eval_list_pop}},
+                {"list.reserve", {&ListReserve::create_ListReserve, &ListReserve::eval_list_reserve}},
```
Comment (czgdp1807) [inline src/libasr/pass/intrinsic_function_registry.h:3345]:
```suggestion
                {"list.reserve", {&ListReserve::create_Reserve, &ListReserve::eval_reserve}},
```

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Comment by certik** [comment]
Beautiful. Yes, as Gagandeep said, you just do:
```
from lpython import reserve
reserve(l, 100)
```
Let's get this in and benchmark again.

**Comment by certik** [comment]
@czgdp1807 if you could please help, that would be great. Let's get this finished.

**Comment by czgdp1807** [comment]
I will meet @kabra1110 today after our meeting ends. We will complete it then.

**Comment by certik** [comment]
Thanks @czgdp1807 ! 

How do the updated benchmarks look like?

**Pull request merged**